src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,403 @@
+	 * modulation pedal inputs: The Portamento, and Sustain pedals are mapped to 
+	 * bits 6, and 7 respectively.
+	 * Note: Input lines 4-15 are used to map the keyboard, which is not implemented here.
+	 * When the keyboard state is read, the default value of 0 will be returned.
+	 * @param offset The offset into the memory mapped region being read.
+	 * @return uint8_t The value read from the bus.
+	 */
+	uint8_t key_switch_scan_driver_r(offs_t offset)
+	{
+		return m_key_switch_input[m_key_switch_input_select]->read();
+	}
+
+	/**
+	 * @brief Handles a write to the 7-segment LED memory mapped region.
+	 * This function is responsible for setting the two 7-segment LEDs set in the
+	 * device's layout file.
+	 * @param offset The offset into the memory mapped region being written.
+	 * @param data The data being written.
+	 */
+	void led_w(offs_t offset, uint8_t data);
+
+	void palette_init(palette_device &palette);
+
+	void mem_map(address_map &map);
+
+	WRITE_LINE_MEMBER(midi_r) { m_rx_data = state; }
+
+	WRITE_LINE_MEMBER(midiclock_w) { if (state) m_maincpu->m6801_clock_serial(); }
+
+	/**
+	 * @brief Handle a write to the synth's IO Port 1.
+	 * IO Port 1 is mapped as follows:
+	 *  Bit 0: Keyboard/Switch Scan Driver Input.
+	 *  Bit 1: "".
+	 *  Bit 2: "".
+	 *  Bit 3: "".
+	 *  Bit 4: ADC EOC Input Line.
+	 *  Bit 5: Cassette Interface Remote Port.
+	 *  Bit 6: Cassette Interface Tape Output.
+	 *  Bit 7: Cassette Interface Tape Input.
+	 * @param offset The offset into the memory mapped region being written.
+	 * @param data The data being written.
+	 */
+	void p1_w(offs_t offset, uint8_t data);
+
+	/**
+	 * @brief Handle a read from the synth's IO Port 1.
+	 * @param offset The offset into the memory mapped region being read.
+	 * @return uint8_t The value read from the port.
+	 */
+	uint8_t p1_r(offs_t offset);
+};
+
+
+/**
+ * yamaha_dx9_state::machine_start
+ */
+void yamaha_dx9_state::machine_start()
+{
+	m_leds.resolve();
+	m_rx_data = ASSERT_LINE;
+}
+
+
+/**
+ * yamaha_dx9_state::lcd_pixel_update
+ */
+HD44780_PIXEL_UPDATE(yamaha_dx9_state::lcd_pixel_update)
+{
+	if (x < 5 && y < 8 && line < 2 && pos < 16)
+			bitmap.pix(line * 10 + y + 1 + ((y == 7) ? 1 : 0), pos * 6 + x + 1) = state ? 1 : 2;
+}
+
+
+/**
+ * yamaha_dx9_state::palette_init
+ */
+void yamaha_dx9_state::palette_init(palette_device &palette)
+{
+	palette.set_pen_color(0, rgb_t(0x87, 0xAD, 0x34)); // background
+	palette.set_pen_color(1, rgb_t(0x0, 0x0, 0x0)); // lcd pixel on
+	palette.set_pen_color(2, rgb_t(0x7D, 0x9F, 0x32)); // lcd pixel off
+}
+
+
+/**
+ * yamaha_dx9_state::mem_map
+ */
+void yamaha_dx9_state::mem_map(address_map &map)
+{
+	// Internal CPU registers.
+	map(0x0000, 0x001f).m(m_maincpu, FUNC(hd6303r_cpu_device::m6801_io));
+
+	map(0x0020, 0x0020).r(FUNC(yamaha_dx9_state::key_switch_scan_driver_r));
+
+	map(0x0022, 0x0022).r("adc", FUNC(m58990_device::data_r));
+	map(0x0024, 0x0024).w("adc", FUNC(m58990_device::address_data_start_w));
```

Comment:
You can use the object finder `m_adc` here rather than the literal tag – it avoids the possibility of certain classes of error.